File: std/runtime/js/wj_subprocess.mjs
// Windjammer std::subprocess backend for Node.js (child_process).
import { spawn } from "node:child_process";
import readline from "node:readline";

const sessions = new Map();
let nextId = 1;

export function spawnProcess(program, args = []) {
  const child = spawn(program, args, { stdio: ["pipe", "pipe", "inherit"] });
  const id = nextId++;
  const rl = readline.createInterface({ input: child.stdout });
  sessions.set(id, { child, rl });
  return { id };
}

export function writeLine(handle, line) {
  const s = sessions.get(handle.id);
  if (!s) throw new Error("invalid handle");
  s.child.stdin.write(line + "\n");
}

export async function readLine(handle) {
  const s = sessions.get(handle.id);
  if (!s) throw new Error("invalid handle");
  for await (const line of s.rl) return line;
  throw new Error("stdout closed");
}

File: std/runtime/go/wj_subprocess.go
// Windjammer std::subprocess backend for Go (os/exec).
package wjsubprocess

import (
	"bufio"
	"fmt"
	"os/exec"
)

type Handle struct {
	ID int
}

var sessions = map[int]*exec.Cmd{}
var readers = map[int]*bufio.Reader{}
var writers = map[int]exec.StdinPipeWriter{}
var nextID = 1

func Spawn(program string, args []string) (Handle, error) {
	cmd := exec.Command(program, args...)
	stdin, err := cmd.StdinPipe()
	if err != nil {
		return Handle{}, err
	}
	stdout, err := cmd.StdoutPipe()
	if err != nil {
		return Handle{}, err
	}
	if err := cmd.Start(); err != nil {
		return Handle{}, err
	}
	id := nextID
	nextID++
	sessions[id] = cmd
	readers[id] = bufio.NewReader(stdout)
	// store stdin writer when Go codegen wires full MCP support
	_ = stdin
	return Handle{ID: id}, nil
}

func WriteLine(h Handle, line string) error {
	return fmt.Errorf("wjsubprocess.WriteLine: wire via Go codegen import")
}
